x-pack/plugins/security/public/session/session_timeout.tsx

```diff
@@ -6,239 +6,367 @@
-      this.channel.close();
-    }
-    this.cleanup();
+  public stop() {
+    const disabled = {
+      lastExtensionTime: 0,
+      expiresInMs: 0,
+      canBeExtendedByMs: 0,
+    };
+    this.toggleEventHandlers(disabled);
+    this.resetTimers(disabled);
+    this.subscription?.unsubscribe();
   }
 
   /**
-   * When the user makes an authenticated, non-system API call, this function is used to check
-   * and see if the session has been extended.
-   * @param url The URL that was called
+   * Event handler that receives session information from other browser tabs.
+   */
+  private handleChannelMessage = (message: SessionState) => {
+    this.sessionState$.next(message);
+  };
+
+  /**
+   * HTTP request interceptor which ensures that API calls extend the session only if tab is
+   * visible.
    */
-  extend(url: string) {
-    // avoid an additional API calls when the user clicks the button on the session idle timeout
-    if (url.endsWith(SESSION_ROUTE)) {
+  private handleHttpRequest = (fetchOptions: HttpFetchOptionsWithPath) => {
+    // Ignore requests to external URLs
+    if (!fetchOptions.path.startsWith('/')) {
       return;
     }
 
-    const { isLifespanTimeout } = this.getTimeout();
-    if (this.warningToast && !isLifespanTimeout) {
-      // the idle timeout warning is currently showing and the user has clicked elsewhere on the page;
-      // make a new call to get the latest session info
-      return this.fetchSessionInfoAndResetTimers();
+    if (fetchOptions.asSystemRequest === undefined) {
+      return { asSystemRequest: !this.isVisible };
     }
-  }
+  };
 
   /**
-   * Fetch latest session information from the server, and optionally attempt to extend
-   * the session expiration.
+   * HTTP response interceptor which allows us to prevent extending the session manually if it has
+   * already been extended as part of an API call.
    */
-  private fetchSessionInfoAndResetTimers = async (extend = false) => {
-    const method = extend ? 'POST' : 'GET';
-    try {
-      const result = await this.http.fetch(SESSION_ROUTE, { method, asSystemRequest: !extend });
-
-      this.handleSessionInfoAndResetTimers(result);
+  private handleHttpResponse = (httpResponse: HttpResponse) => {
+    // Ignore session endpoint which is already handled by fetch callback
+    if (httpResponse.fetchOptions.path === SESSION_ROUTE) {
+      return;
+    }
 
-      // share this updated session info with any other tabs to sync the UX
-      if (this.channel) {
-        this.channel.postMessage(result);
+    // Only mark session as extended if not a system request
+    if (httpResponse.fetchOptions.asSystemRequest === false) {
+      const currentState = this.sessionState$.getValue();
+      if (currentState.canBeExtendedByMs) {
+        const nextState = {
+          ...currentState,
+          lastExtensionTime: Date.now(),
+          expiresInMs: currentState.canBeExtendedByMs,
+        };
+        this.sessionState$.next(nextState);
```

Comment:
I haven't tested this, but it seems like it could run into issues in some corner cases. It would be difficult to trigger this behavior manually but I think could be checked in unit tests.

Example:
* I configure Kibana with an `idleTimeout` of 2m and a `lifespan` of 5m
* I log in, I wait 59s, and I refresh the page (`idleTimeoutExpiration` is 2m from now, `lifespanExpiration` is 4m1s from now, so `canBeExtendedByMs` is set to 2000 and `expiresInMs` is set to 2m from now)
* I wait 1m59s, and I make an API call (`idleTimeoutExpiration` is 2m from now, `lifespanExpiration` is 2m2s from now, and `expiresInMs` is set to 2m from now)
* I wait 59s, and I make an API call (`idleTimeoutExpiration` is 2m from now, `lifespanExpiration` is 1m3s from now, and `expiresInMs` is set to 2m from now)

All of the client-side timers are calculated based on `expiresInMs`, but they are now wrong. The session will expire 57s before the client thinks it will expire. Of course this is a contrived example but you could imagine it happening with larger values for `idleTimeout` and `idleTimeoutExpiration`.

I think the worst thing we can do here is to allow the session to expire without showing anything to a user -- if your session expires but the page doesn't know it, then click a button and all of a sudden you are kicked back to the login page, that's a frustrating experience.

One option that I can see to handle this edge case is to update the `SessionState` interface and add a `maxExpiresInMs` value that is calculated based on the `lifespanExpiration`. This could be calculated when `SessionInfo` is fetched from the server, and it could be used to ensure we don't set `expiresInMs` to an unrealistic value.

I had to write it out to wrap my head around it, so I'll just paste the diff here in case you want to use this approach:

<details><summary>See diff</summary>

```diff
diff --git a/x-pack/plugins/security/public/session/session_timeout.tsx b/x-pack/plugins/security/public/session/session_timeout.tsx
index 8fe9865854f..fcd72e02397 100644
--- a/x-pack/plugins/security/public/session/session_timeout.tsx
+++ b/x-pack/plugins/security/public/session/session_timeout.tsx
@@ -58,6 +58,7 @@ export interface ISessionTimeout {
 export interface SessionState {
   lastExtensionTime: number;
   expiresInMs: number;
+  maxExpiresInMs: number;
   canBeExtendedByMs: number;
 }
 
@@ -70,6 +71,7 @@ export class SessionTimeout {
   private sessionState$ = new BehaviorSubject<SessionState>({
     lastExtensionTime: 0,
     expiresInMs: 0,
+    maxExpiresInMs: 0,
     canBeExtendedByMs: 0,
   });
   private subscription?: Subscription;
@@ -107,6 +109,7 @@ export class SessionTimeout {
     const disabled = {
       lastExtensionTime: 0,
       expiresInMs: 0,
+      maxExpiresInMs: 0,
       canBeExtendedByMs: 0,
     };
     this.toggleEventHandlers(disabled);
@@ -149,11 +152,16 @@ export class SessionTimeout {
     // Only mark session as extended if not a system request
     if (httpResponse.fetchOptions.asSystemRequest === false) {
       const currentState = this.sessionState$.getValue();
-      if (currentState.canBeExtendedByMs) {
+      const { expiresInMs, maxExpiresInMs, canBeExtendedByMs } = currentState;
+      if (canBeExtendedByMs && (!maxExpiresInMs || maxExpiresInMs > expiresInMs)) {
+        let newExpiresInMs = canBeExtendedByMs;
+        if (maxExpiresInMs && expiresInMs > maxExpiresInMs) {
+          newExpiresInMs = maxExpiresInMs;
+        }
         const nextState = {
           ...currentState,
           lastExtensionTime: Date.now(),
-          expiresInMs: currentState.canBeExtendedByMs,
+          expiresInMs: newExpiresInMs,
         };
         this.sessionState$.next(nextState);
         if (this.channel) {
@@ -268,12 +276,16 @@ export class SessionTimeout {
       });
 
       let expiresInMs = 0;
+      let maxExpiresInMs = 0;
       let canBeExtendedByMs = sessionInfo.idleTimeout || 0;
 
       const { now, idleTimeoutExpiration, lifespanExpiration } = sessionInfo;
       if (idleTimeoutExpiration) {
         expiresInMs = idleTimeoutExpiration - now;
       }
+      if (lifespanExpiration) {
+        maxExpiresInMs = lifespanExpiration - now;
+      }
       if (
         lifespanExpiration &&
         (idleTimeoutExpiration === null || lifespanExpiration <= idleTimeoutExpiration)
@@ -285,6 +297,7 @@ export class SessionTimeout {
       const nextState = {
         lastExtensionTime: Date.now(),
         expiresInMs,
+        maxExpiresInMs,
         canBeExtendedByMs,
       };
       this.sessionState$.next(nextState);
```

</details>

Reply:
Good catch! There could be an issue when both, `idleTimeout` and `lifespan` are configured, where the session might get extended beyond its lifespan. 

We can change this to fetch session information from the server after intercepting an HTTP response rather than "blindly" incrementing it client side. However, that would essentially double the HTTP requests the front-end makes which I'd prefer to avoid. 

Is there a third option where any API call that extends the session returns that information as a custom header in its response so that the client can read that out instead of manually polling?

Reply:
I'm sure it's possible to tack on an additional HTTP response header but it sounds a bit wasteful IMO, not sure how the Core team would feel about it.

I don't have stats in front of me but I'm pretty confident that the vast majority of users won't have an idleTimeout less than 30 minutes ([that is what NIST recommends](https://pages.nist.gov/800-63-3/sp800-63b.html)). As I mentioned in https://github.com/elastic/kibana/pull/98461#pullrequestreview-647043231, I don't think we need to update the client's session info with every request -- we already have a refresh timer set to kick off right before we show the warning.

Since the toast is shown 5 minutes before the session expires, assuming a 30 minute idle timeout, we are fetching session info once every 25 minutes. Or, to rephrase: this bit in the response interceptor to extend the client-side timers is really only saving us one HTTP request every 25 minutes. I don't think it's worth the additional complexity.

Reply:
The number of API calls we would introduce to the extend session endpoint is higher than that unfortunately. It’s defined by `EXTENSION_THROTTLE_MS` which I have set to 1 minute. So it would currently be 1 extra API call per minute for every single user. That's probably totally acceptable but want to make sure we're not adding unnecessary traffic.

Reply:
> The number of API calls we would introduce to the extend session endpoint is higher than that unfortunately. It’s defined by `EXTENSION_THROTTLE_MS` which I have set to 1 minute. So it would currently be 1 extra API call per minute for every single user. That's probably totally acceptable but want to make sure we're not adding unnecessary traffic.

I think an extra call per user per minute is acceptable. The amount of work we do for this particular request is minimal comparatively, and I don't foresee this being problematic in production setups.